### PR TITLE
The working version

### PR DESCRIPTION
**Pull request by maxim-fedotov**

### CODE REVIEW

File: latin_squares.py [hunk 1]
```diff
@@ -133,226 +129,196 @@ def resample(self):  # TODO
         print("Done!")
         return
 
-    def get_lambda(self):
+    def ess_form(self, delta):  # DONE
+        V = np.array([self.V(p) for p in self.particles])
+        w = np.exp(- delta * V)
+        return np.sum(w)**2 / np.sum(w**2)
```
Comment (antotocar34) [inline latin_squares.py:135]: Might be more efficient to let
`w = np.exp( - 2 * delta * V)`

File: latin_squares.py [hunk 2]
```diff
@@ -133,226 +129,196 @@ def resample(self):  # TODO
...
         try:
-            delta = root_scalar(
-                f, 
-                bracket=[0, self.lambda_max - self.lambdas[-1]],
-                method="brentq"
-            ).root  # Not sure about this bracket argument
-        except ValueError: # <- If a solution is not found (see algorithm 17.3 in book)
-            print("Delta does not exist")
-            delta = self.lambda_max - self.lambdas[-1]
+            delta = root_scalar(lambda d: self.ess_form(d) - self.ess_min,
```
Comment (antotocar34) [inline latin_squares.py:146]: lovely, much cleaner :D

File: latin_squares.py [hunk 3]
```diff
@@ -133,226 +129,196 @@ def resample(self):  # TODO
...
-        sum(
-            (
-                sum((x[i, j] == 1) for i in range(d)) ** 2
-                for l in range(d)
-                for j in range(d)
-            )
-        )
-        - d ** 2
-    )
+    return sum(sum(sum(x[i, j] == l for i in range(d))**2 for l in range(d)) - d for j in range(d))
```
Comment (antotocar34) [inline latin_squares.py:236]: perhaps
`
sum(sum(sum(x[i, j] == l for i in range(d))**2 for l in range(d)) for j in range(d)) - d**2 is better? or maybe it doesn't make a difference idk
`

File: latin_squares.py (hunk 1 above)
Comment (maxim-fedotov) [reply latin_squares.py:135]: yeah, feel free to change

Comment (antotocar34) [reply latin_squares.py:135]: I think it's easier for you to do it before I pull your changes in. Sorry ;)

File: latin_squares.py (hunk 3 above)
Comment (maxim-fedotov) [reply latin_squares.py:236]: yeah, might be, idk whether it gives any computational gain

File: latin_squares.py (hunk 1 above)
Comment (maxim-fedotov) [reply latin_squares.py:135]: Yeah, sure

**Pull request merged**